File: src/extract_videos.cpp
#include "../include/make_directory.hpp"

class Extract_videos {
private:
  fs::path input_path;
  fs::path images_path;
  fs::path flow_x_path;
  fs::path flow_y_path;
  fs::path output_path;
  fs::path flow_type;
  string log_path;
  string image_name;
  int image_create_flag;
  int flow_type_id;
public:
  void large_scale_data_processing() {
    fs::path images_path, flow_x_path, flow_y_path, input_path;
    fs::path output_path_full, video_path;
    string output_path_full_string, video_path_string;
    ofstream log_file(this->log_path, ios::app);
    int count = 0;

    if (log_file.is_open() == 0) {
      cout << "Failed to open the file :" << this->log_path << endl;
      exit(1);
    }
    if (this->image_create_flag == 0)   {
      make_directory(this->input_path, this->images_path);
    }
    make_directory(this->input_path, this->flow_x_path);
    make_directory(this->input_path, this->flow_y_path);

    for (auto & class_dir : fs::directory_iterator(this->input_path)) {
      for (auto & video : fs::directory_iterator(class_dir)) {
        video_path = fs::path(video);
        video_path = fs::path(video_path.string().substr(string(this->input_path).length()));
        images_path = this->images_path / video_path;
        flow_x_path = this->flow_x_path / video_path;
        flow_y_path = this->flow_y_path / video_path;
        input_path = this->input_path / video_path;
        if (fs::is_empty(flow_x_path) or fs::is_empty(flow_y_path)) {
          system(("extract_gpu -f=" + input_path.string() + " -i=" + images_path.string() + this->image_name + " -x=" + flow_x_path.string() + this->image_name + " -y=" + flow_y_path.string() + this->image_name + " -o=dir -t=" + to_string(this->flow_type_id) + " -n=" + to_string(this->image_create_flag)).c_str());
          log_file << video_path.string() << endl; 
          cout << ++count << " " << video_path.string() << endl;
        }
      }
    }
    log_file.close();
  }

  Extract_videos(char** argv) {
    this->flow_type_id = stoi(string(argv[4]));
    switch(this->flow_type_id) {
    case 0 :
      flow_type = fs::path("flow/farn");
      break;
    case 1 :
      flow_type = fs::path("flow/tvl1");
      break;
    case 2 :
      flow_type = fs::path("flow/brox");
    }
    this->image_name = "/image";
    fs::path pwd = string(getenv("PWD"));
    this->image_create_flag = stoi(string(argv[5]));
    this->output_path = pwd / fs::path(string(argv[2]));
    this->images_path = output_path / fs::path("images");
    this->flow_x_path = output_path / flow_type / fs::path("flow_x");
    this->flow_y_path = output_path / flow_type / fs::path("flow_y");
    this->input_path = pwd / fs::path(string(argv[1]));
    this->log_path = string(argv[3]);
  }
};

// input path, output path, log_path, optical flow type, image create flag
int main(int argc, char** argv) {
  Extract_videos *extract_videos = new Extract_videos(argv);
  extract_videos->large_scale_data_processing();
  return 0;
}

File: src/extract_videos.py
import argparse
import os
import re
import time

from joblib import Parallel, delayed


def parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        'input_path', default=None, type=str, help='input path'
    )
    parser.add_argument(
        'output_path', default=None, type=str, help='output path'
    )
    parser.add_argument(
        '--image_frag', default=0, type=int, help='create image flag'
    )
    parser.add_argument(
        '--flow_type', default=0, type=int, help='flow type id'
    )
    parser.add_argument(
        '--fps', default=1, type=int, help='frame rate'
    )
    return parser.parse_args()


class Extract_videos():
    def __init__(self, args):
        self.args = args
        self.flow = {0: 'farn', 1: 'tvl1', 2: 'brox'}
        self.input_video_paths = []
        self.images_paths = []
        self.flow_x_paths = []
        self.flow_y_paths = []
        self.count = 1

    def make_directory(self):
        input_path = self.args.input_path
        output_path = self.args.output_path
        for class_dir in os.listdir(input_path):
            class_dir_full = os.path.join(input_path, class_dir)
            for video in os.listdir(class_dir_full):
                video_path = os.path.join(input_path, class_dir, video)
                video_name = re.sub(r'\.(avi|mkv|mp4|webm)', '', video)
                image_path = os.path.join(output_path, 'images', class_dir, video_name)
                flow_x_path = os.path.join(output_path, 'flow', self.flow[self.args.flow_type], 'flow_x', class_dir,
                                           video_name)
                flow_y_path = os.path.join(output_path, 'flow', self.flow[self.args.flow_type], 'flow_y', class_dir,
                                           video_name)
                if not os.path.isfile(os.path.join(flow_x_path, 'image_00001.jpg')):
                    os.makedirs(image_path, exist_ok=True)
                    os.makedirs(flow_x_path, exist_ok=True)
                    os.makedirs(flow_y_path, exist_ok=True)
                    self.input_video_paths.append(video_path)
                    self.images_paths.append(image_path)
                    self.flow_x_paths.append(flow_x_path)
                    self.flow_y_paths.append(flow_y_path)
                    print('create {}'.format(flow_y_path))

    def process(self, input_video_path, images_path, flow_x_path, flow_y_path):
        file_name = '/image'
        regex = r'([()&;])'
        regex_later = r'\\\1'
        input_video_path = re.sub(regex, regex_later, input_video_path)
        images_path = re.sub(regex, regex_later, images_path)
        flow_x_path = re.sub(regex, regex_later, flow_x_path)
        flow_y_path = re.sub(regex, regex_later, flow_y_path)
        os.system(
            'extract_gpu -f={} -i={} -x={} -y={} -n={} -t={} -o=dir -s={}'.format(input_video_path,
                                                                            images_path + file_name,
                                                                            flow_x_path + file_name,
                                                                            flow_y_path + file_name,
                                                                            self.args.image_frag, self.args.flow_type, self.args.fps))
        print(input_video_path)

    def large_data_processing(self):
        Parallel(n_jobs=-1)(
            [delayed(self.process)(self.input_video_paths[i], self.images_paths[i], self.flow_x_paths[i],
                                   self.flow_y_paths[i]) for i in range(len(self.input_video_paths))])


if __name__ == '__main__':
    extract_videos = Extract_videos(parser())
    extract_videos.make_directory()
    start_time = time.time()
    extract_videos.large_data_processing()
    end_time = time.time() - start_time
    print('Time :{}[sec]'.format(end_time))
